doc/neps/nep-0037-array-module.rst

```diff
@@ -0,0 +1,332 @@
+
+            # Traverse submodules to find the appropriate function
+            modules = func.__module__.split('.')
+            assert modules[0] == 'numpy'
+            for submodule in modules[1:]:
+                module = getattr(module, submodule, None)
+            new_func = getattr(module, func.__name__, None)
+            if new_func is None:
+                return NotImplemented
+
+            return new_func(*args, **kwargs)
+
+Alternatives considered
+-----------------------
+
+Naming
+======
+
+We like the name ``__array_module__`` because it mirrors the existing ``__array_function__`` and ``__array_ufunc__`` protocols. Another reasonable choice could be ``__array_namespace__``.
+
+It is less clear what the NumPy function that calls this protocol should be called (``get_array_module`` in this proposal). Some possible alternatives: ``array_module``, ``common_array_module``, ``resolve_array_module``, ``get_namespace``, ``get_numpy``, ``get_numpylike_module``, ``get_duck_array_module``.
+
+Requesting restricted subsets of NumPy's API
+============================================
+
+Over time, NumPy has accumulated a very large API surface, with over 600 attributes in the top level ``numpy`` module alone. It is unlikely that any duck array library could or would want to implement all of these functions and classes, because the frequently used subset of NumPy is much smaller.
+
+We think it would be useful exercise to define "minimal" subset(s) of NumPy's API, omitting rarely used or non-recommended functionality. For example, minimal NumPy might include ``stack``, but not the other stacking funtions ``column_stack``, ``dstack``, ``hstack`` and ``vstack``. This could clearly indicate to duck array authors and users want functionality is core and what functionality they can skip.
+
+Support for requesting a restricted subset of NumPy's API would be a natural feature to include in  ``get_array_function`` and ``__array_module__``, e.g.,
+
+.. code:: python
+
+    # array_module is only guranteed to contain "minimal" NumPy
+    array_module = np.get_array_module(*arrays, request='minimal')
+
+To facilitate testing with NumPy and use with any valid duck array library, NumPy itself would return restricted versions of the ``numpy`` module when ``get_array_module`` is called only on NumPy arrays. Omited functions would simply not exist.
+
+Unfortuntely, we have not yet figured out what these restricted subsets should be, so it doesn't make sense to do this yet. When/if we do, we could either add new keyword arguments to ``get_array_module`` or add new top level functions, e.g., ``get_minimal_array_module``. We would also need to add a new protocol patterned off of ``__array_module__`` (e.g., ``__array_module_minimal__``), because passing new arguments into methods is not backwards compatible. (We could anticipate this need by adding placeholder arguments into  ``__array_module__``, but it probably isn't worth the trouble.)
```

Comment:
It might be possible to just say that if the keyword argument is not accepted, `"full"` is assumed to be supported. If we want to transition there, it would require a `try/except` clause at some point (giving a deprecation warning when passing the kwarg fails). That probably is viable...